### PR TITLE
Don't pattern match on all arguments unless needed

### PR DESCRIPTION
**Pull request by rxin**
I've seen this all over Spark, and it is a pain when reviewing code to make sure things align properly, and a big pain in refactoring.

### CODE REVIEW

**Comment by rxin** [comment]
cc @marmbrus

File: README.md [hunk 1]
```diff
@@ -365,6 +365,32 @@ Of course, the situation in which a class grows this long is strongly discourage
   }
   ```
 
+- If the only goal is to match on the type of the object, do NOT expand fully all the arguments, as it makes refactoring more difficult and the code more error prone.
```
Comment (marmbrus) [inline README.md]: I agree with caution here, especially in a case like this where there are a bunch of back to back integers (you might even call that out.  I think its the same thing as using named arguments).  Maybe this should more be pro/cons of use extractors vs. matching on types:

Pro-extactors:
  - few arguments
  - nested matching

Avoid extractors:
  - Only using some of the arguments
  - Many arguments of the same type

Comment (aarondav) [inline README.md]: The latter is pretty hard to reason about, are there easier heuristics available?

Comment (rxin) [inline README.md]: "classes that are well-defined and do not expect to be extended"

No class should have this assumption / expectation :)

I like the idea of having a separate extractor vs no-extractors section.

Comment (aarondav) [inline README.md]: e.g., `WorkspaceId(workspaceId)` does not expect to be extended. Or `DbAwsKeys(accessKey, secretKey)`. They are fully self-contained concepts that would be as soon replaced as extended.

Comment (rxin) [inline README.md]: I don't know if that's true. One day we might change those too.

Comment (rxin) [inline README.md]: Also even if that was true, it's in general very difficult for engineers to judge the future evolution of code.

Comment (aarondav) [inline README.md]: Which is why extractors should be used sparingly!

Comment (marmbrus) [inline README.md]: For something like `WorkspaceId`, if you add something (e.g. `orgId`), I'd actually want compile errors all over the place to go and handle that.

Comment (aarondav) [inline README.md]: Can you come up with criteria other than "nested matching" then where it is good to use extractors? NonFatal for example.

Comment (marmbrus) [inline README.md]: If you are matching and using all the argument, it seems likely that adding a new argument should have you revisit that code to see if it needs to be handled.  I'm not sure if its a "pro" but its at least acceptable.

Comment (rxin) [inline README.md]: Do you guys have objections about merging this as is?

**Comment by aarondav** [comment]
Merging because the "If" clause at the beginning seems to be in line with our discussion -- there are use-cases for extractors, but not "if the only goal is to match on the type of the object".

**Comment by rxin** [comment]
cc @HyukjinKwon and @Hawstein on this one too.

**Pull request merged**